patroni/postgresql.py

```diff
@@ -655,12 +750,84 @@ def reload(self):
             self.call_nowait(ACTION_ON_RELOAD)
         return ret
 
-    def restart(self):
-        self.set_state('restarting')
-        ret = self.stop(block_callbacks=True) and self.start(block_callbacks=True)
-        if ret:
-            self.call_nowait(ACTION_ON_RESTART)
+    def check_for_startup(self):
+        """Checks PostgreSQL status and returns if PostgreSQL is in the middle of startup."""
+        if self.is_starting():
+            return not self.check_startup_state_changed()
         else:
+            ready = self.pg_isready()
```

Comment:
Yeah, that would in principle work. Both `Ha.follow()` and `Ha.enforce_master_role()` call `Postgresql.is_leader()` and from what I can tell all relevant code paths end up there or `Postgresql.stop()` (via `demote()` or `restart()`). `is_leader()` should then catch connection errors and then check pg_isready that the server is indeed starting up, psycopg2 doesn't allow us to distinguish between server down and server starting up without parsing possibly localized error messages.

That said, I don't see a huge issue with checking for pg_isready on each iteration, but I do find loop-to-loop state propagation confusing and hard to understand as it is. Having something  called is_leader() change state to signal some other code about state change starts to get a bit too "spooky action from a distance" for my taste.

I'd actually like to refactor the code a bit to make state modification much more explicit and most of the code purely functional. The main idea is to have the main function take as input cluster state from DCS and current PostgreSQL state and have that output the desired action to take. But this idea is in pretty early phases at this point.